src/call/mod.rs

```diff
@@ -683,4 +711,46 @@ impl SinkBase {
         self.batch_f.take();
         Poll::Ready(Ok(()))
     }
+
+    #[inline]
+    fn poll_flush<C: ShareCallHolder>(
+        &mut self,
+        cx: &mut Context,
+        call: &mut C,
+    ) -> Poll<Result<()>> {
+        if self.batch_f.is_some() {
+            ready!(self.poll_ready(cx)?);
+        }
+        if self.buf_flags.is_some() {
+            self.start_send_buffer_message(self.last_buf_hint, call)?;
+            ready!(self.poll_ready(cx)?);
+        }
+        Poll::Ready(Ok(()))
+    }
+
+    #[inline]
+    fn start_send_buffer_message<C: ShareCallHolder>(
+        &mut self,
+        buffer_hint: bool,
+        call: &mut C,
+    ) -> Result<()> {
+        // `start_send` is supposed to be called after `poll_ready` returns ready.
+        assert!(self.batch_f.is_none());
+
+        let mut flags = self.buf_flags.clone().unwrap();
+        flags = flags.buffer_hint(buffer_hint);
+        let write_f = call.call(|c| {
+            c.call
+                .start_send_message(&self.buffer, flags.flags, self.send_metadata)
+        })?;
+        self.batch_f = Some(write_f);
+        self.buf_flags.take();
+        // NOTE: Content of `self.buf` is copied into grpc internal.
+        self.buffer.clear();
+        if self.buffer.capacity() > BUF_SHRINK_SIZE {
+            self.buffer.truncate(BUF_SHRINK_SIZE);
```

Comment:
Because it's cleared at L749, so truncate will do nothing, and `shrink_to_fit` may shrink it to 0.

Reply:
Yes, I removed `truncate` and if it grows up to `BUF_SHRINK_SIZE`, shrink can do some cleanup.

Reply:
This is wrong. The whole point of buffer is to reduce allocation, this implement will make it allocate more.